core/valor_core/detection.py

```diff
@@ -0,0 +1,1470 @@
+        )
+    )
+
+    return joint_df
+
+
+def _get_dtypes_in_series_of_arrays(series: pd.Series):
+    """Get the data type inside of a 2D numpy array. Used to check if a np.array contains coordinates or a mask."""
+    if not isinstance(series, pd.Series) or not all(
+        series.apply(lambda x: x.ndim == 2)
+    ):
+        raise ValueError(
+            "series must be a pandas Series filled with two-dimensional arrays."
+        )
+
+    unique_primitives = series.map(lambda x: x.dtype).unique()
+
+    if len(unique_primitives) > 1:
+        raise ValueError("series contains more than one type of primitive.")
+
+    return unique_primitives[0]
+
+
+def _check_if_series_contains_masks(series: pd.Series) -> bool:
+    """Check if any element in a pandas.Series is a mask."""
+    if series.empty:
+        return False
+
+    primitive = _get_dtypes_in_series_of_arrays(series=series)
+
+    if np.issubdtype(primitive, np.bool_):
+        return True
+
+    return False
+
+
+def _calculate_iou(
+    joint_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """Calculate the IOUs between predictions and groundtruths in a joint dataframe."""
+    if _check_if_series_contains_masks(
+        joint_df.loc[
+            joint_df["converted_geometry_pd"].notnull(),
+            "converted_geometry_pd",
+        ]
+    ):
+        iou_calculation_df = (
+            joint_df.assign(
+                intersection=lambda chain_df: chain_df.apply(
+                    lambda row: (
+                        0
+                        if row["converted_geometry_pd"] is None
+                        or row["converted_geometry_gt"] is None
+                        else np.logical_and(
+                            row["converted_geometry_pd"],
+                            row["converted_geometry_gt"],
+                        ).sum()
+                    ),
+                    axis=1,
+                )
+            )
+            .assign(
+                union_=lambda chain_df: chain_df.apply(
+                    lambda row: (
+                        0
+                        if row["converted_geometry_pd"] is None
+                        or row["converted_geometry_gt"] is None
+                        else np.sum(row["converted_geometry_gt"])
+                        + np.sum(row["converted_geometry_pd"])
+                        - row["intersection"]
+                    ),
+                    axis=1,
+                )
+            )
+            .assign(
+                iou_=lambda chain_df: chain_df["intersection"]
+                / chain_df["union_"]
+            )
+        )
+
+        joint_df = joint_df.join(iou_calculation_df["iou_"])
+
+    else:
+        iou_calculation_df = joint_df.loc[
+            ~joint_df["converted_geometry_gt"].isnull()
+            & ~joint_df["converted_geometry_pd"].isnull(),
+            ["converted_geometry_gt", "converted_geometry_pd"],
+        ].apply(
+            lambda row: geometry.calculate_bbox_iou(
```

Comment:
@ekorman since we are losing the ability to compute IoU for polygons could we reconsider a library like [shapely](https://shapely.readthedocs.io/en/stable/manual.html)?

Reply:
this will be added in a future PR; let's discuss it then (see 'Issues to Fix' in PR notes)